checker/tests/resourceleak-arrays/patternMatchOwningArrayLoops.java

```diff
@@ -0,0 +1,258 @@
+  public void illegalAliasing() {
+    @OwningArray Socket[] arr = new Socket[n];
+    // :: error: illegal.owningarray.assignment
+    @OwningArray Socket[] arr2 = arr;
+    // :: error: illegal.aliasing
+    Socket[] arr3 = arr;
+  }
+
+  // test that declaring an @OwningArray is alright
+  public void illegalOwningArrayElementAssignment() {
+    @OwningArray Socket[] arr = new Socket[n];
+    try {
+      // :: error: illegal.owningarray.element.assignment
+      // :: error: required.method.not.called
+      arr[0] = new Socket(myHost, myPort);
+    } catch (Exception e) {
+    }
+    // :: error: illegal.owningarray.element.assignment
+    arr[0] = null;
+  }
+
+  public void unfulfilledAllocationLoop() {
+    // :: error: unfulfilled.mustcallonelements.obligations
+    @OwningArray Socket[] arr = new Socket[n];
+    for (int i = 0; i < n; i++) {
+      try {
+        arr[i] = new Socket(myHost, myPort);
+      } catch (Exception e) {
+      }
+    }
+  }
+
+  public void reassignOwningArrayWithOpenObligations() {
+    // try to trick the checker by closing the elements of the array
+    // after reassigning it.
+    // :: error: unfulfilled.mustcallonelements.obligations
+    @OwningArray Socket[] arr = new Socket[n];
+    for (int i = 0; i < n; i++) {
+      try {
+        arr[i] = new Socket(myHost, myPort);
+      } catch (Exception e) {
+      }
+    }
+    arr = new Socket[n];
+    for (int i = 0; i < n; i++) {
+      try {
+        arr[i].close();
+      } catch (Exception e) {
+      }
+    }
+  }
+
+  // test that opening and subsequent closing is alright
+  public void validClosing() {
+    @OwningArray Socket[] arr = new Socket[n];
+    for (int i = 0; i < n; i++) {
+      try {
+        arr[i] = new Socket(myHost, myPort);
+      } catch (Exception e) {
+      }
+    }
+    for (int i = 0; i < n; i++) {
+      try {
+        arr[i].close();
+      } catch (Exception e) {
+      }
+    }
+  }
+
+  public void validDeallocationLoop() {
+    @OwningArray Socket[] arr = new Socket[n];
+    for (int i = 0; i < n; i++) {
+      try {
+        arr[i] = new Socket(myHost, myPort);
+      } catch (Exception e) {
+      }
+    }
+    // this deallocation loop is legal and should be pattern-matched
+    for (int i = 0; i < n; i++) {
+      try {
+        try {
+          arr[i].close();
+        } catch (Exception e) {
+        }
+      } catch (Exception e) {
+      }
+    }
+  }
+
+  public void invalidDeallocationLoop() {
+    // :: error: unfulfilled.mustcallonelements.obligations
+    @OwningArray Socket[] arr = new Socket[n];
+    for (int i = 0; i < n; i++) {
+      try {
+        arr[i] = new Socket(myHost, myPort);
+      } catch (Exception e) {
+      }
+    }
+    // this deallocation loop is illegal and is not pattern-matched
+    for (int i = 0; i < n; i++) {
+      try {
+        arr[i].close();
+        i++;
+      } catch (Exception e) {
+      }
+    }
+  }
+
+  public void invalidDeallocationLoop2() {
```

Comment:
This case might be solvable with some integration with the Index Checker. This is explicitly a suggestion for future work, but we might consider including a comment here suggesting that it would be nice to handle cases like this, but that integration with the Index Checker would be necessary.